kernel/src/scan/mod.rs

```diff
@@ -264,15 +267,17 @@ impl Scan {
 
         let global_state = Arc::new(self.global_scan_state());
         let scan_data = self.scan_data(engine)?;
-        let mut scan_files = vec![];
-        for data in scan_data {
-            let (data, vec) = data?;
-            scan_files =
-                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)?;
-        }
-        scan_files
-            .into_iter()
-            .map(|scan_file| -> DeltaResult<_> {
+        let scan_files_iter = scan_data
+            .map_ok(|(data, vec)| {
+                let scan_files = vec![];
+                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)
+                    .map(IntoIterator::into_iter)
```

Comment:
was thinking about whether or not we could 'push down' the iterator here so that we don't have to materialize everything into a vector. but since the visitor eagerly extracts, probably best we can do is this for now?

Reply:
Yep, ScanFileVisitor iterates through all the rows eagerly, and calls the callback in the loop. I could make an issue to make that lazy as well?

Reply:
yea let's have an issue to at least consider it :)

Reply:
This makes sense to me. Probably okay to just `collect` these up since we're just iterating over some in memory data and not doing anything expensive.